From 4b67d5fd4ec1abbf8099c52c2a118e765490ce35 Mon Sep 17 00:00:00 2001
From: Randy Barlow <randy@electronsweatshop.com>
Date: Wed, 21 Sep 2016 18:19:50 -0400
Subject: [PATCH] Do not allow NULL values to be stored in the Comment.text
 column.

This commit disallows the Comment.text column to contain NULL
values. Additionally, it contains a data migration to convert all
NULL values to the empty string.

fixes #949
---
 alembic/versions/37f38ddc4c8d_.py | 32 ++++++++++++++++++++++++++++++++
 bodhi/server/models/models.py     |  2 +-
 2 files changed, 33 insertions(+), 1 deletion(-)
 create mode 100644 alembic/versions/37f38ddc4c8d_.py

diff --git a/alembic/versions/37f38ddc4c8d_.py b/alembic/versions/37f38ddc4c8d_.py
new file mode 100644
index 0000000..824808a
--- /dev/null
+++ b/alembic/versions/37f38ddc4c8d_.py
@@ -0,0 +1,32 @@
+"""Do not allow NULL values in the text column of the comments table.
+
+Revision ID: 37f38ddc4c8d
+Revises: 4df1fcd59050
+Create Date: 2016-09-21 19:51:04.946521
+
+"""
+
+from alembic import op
+import sqlalchemy as sa
+
+
+# revision identifiers, used by Alembic.
+revision = '37f38ddc4c8d'
+down_revision = '4df1fcd59050'
+
+
+def upgrade():
+    """
+    We will need to set all existing NULL comments to "", then change the column to disallow NULL comments.
+    """
+    # Build a fake mini version of the comments table so we can form an UPDATE statement.
+    comments = sa.sql.table('comments', sa.sql.column('text', sa.UnicodeText))
+    # Set existing NULL comments to "".
+    op.execute(comments.update().where(comments.c.text==None).values({'text': op.inline_literal('')}))
+
+    # Disallow new NULL comments.
+    op.alter_column('comments', 'text', existing_type=sa.TEXT(), nullable=False)
+
+
+def downgrade():
+    op.alter_column('comments', 'text', existing_type=sa.TEXT(), nullable=True)
diff --git a/bodhi/server/models/models.py b/bodhi/server/models/models.py
index f096d23..df34891 100644
--- a/bodhi/server/models/models.py
+++ b/bodhi/server/models/models.py
@@ -1904,7 +1904,7 @@ class Comment(Base):
 
     karma = Column(Integer, default=0)
     karma_critpath = Column(Integer, default=0)
-    text = Column(UnicodeText)
+    text = Column(UnicodeText, nullable=False)
     anonymous = Column(Boolean, default=False)
     timestamp = Column(DateTime, default=datetime.utcnow)
 
-- 
2.10.0

